### PR TITLE
feat: step events frontend

### PR DESCRIPTION
**Pull request by csiyang**
# Description

### <samp>🤖 Generated by Copilot at dc4bab4</samp>

This pull request adds a new feature to track and create events when users navigate to the previous step in a journey. It updates the GraphQL schemas, resolvers, and types of the api-gateway, api-journeys, and journeys-ui apps to support the new mutation field `stepPrevEventCreate`. It also adds the corresponding UI components, tests, and data layer events to the journeys and journeys-admin apps. It modifies the Event database table and model to include a new column and field `prevStepId`. It fixes some typos and adds some comments to improve readability.

[- Link to Basecamp Todo](https://3.basecamp.com/3105655/buckets/34951357/todos/6733851162)

Requires: https://github.com/JesusFilm/core/pull/2022

# How should this PR be QA Tested?

Please describe the QA tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

# Walkthrough

### <samp>🤖 Generated by Copilot at dc4bab4</samp>

*  Add stepPrevEventCreate mutation field to create and track events when users navigate to the previous step in a journey ([link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-e60078ed0fc71b8cd3bb2fca25505e7dba866b7d463147662a12432dfe287273R1148), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3cbd0d77b8eff95fb0f62db7f9f1a4b69ccd568ae3902a3450bac7d752c817a3R227), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-98be3b0df88f6c60d0542c483b910caa761b1470bb6c87461e9b3de7f98e63b5R18), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-98be3b0df88f6c60d0542c483b910caa761b1470bb6c87461e9b3de7f98e63b5R47), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-29131681a98f41397b30e90886c80b6fc28d42df8c3c5b7df71856a044d9b5abL73-R125), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-b62940f4fc34e8998562c7e78b1e73dde6b81d243daa3516e0ce387838766d21R12-R13), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-b62940f4fc34e8998562c7e78b1e73dde6b81d243daa3516e0ce387838766d21R117-R145), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-684ea1b2346c27f9252ea686988ffaa756a41f56bd7c62530441bfe2d034df3dR1), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-684ea1b2346c27f9252ea686988ffaa756a41f56bd7c62530441bfe2d034df3dL8-R21), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-684ea1b2346c27f9252ea686988ffaa756a41f56bd7c62530441bfe2d034df3dL35-R56), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-684ea1b2346c27f9252ea686988ffaa756a41f56bd7c62530441bfe2d034df3dL73-R154), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-98c806310428fde902045482ce23027cd6ab7578d9e4048de778be28eed4b4caR200-R215), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-83b79888786fd47915f01e3ae6fe2d2fb314332f9ef05f5fa8168dfff174f6cfL1-R11), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-83b79888786fd47915f01e3ae6fe2d2fb314332f9ef05f5fa8168dfff174f6cfL16-R41), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-83b79888786fd47915f01e3ae6fe2d2fb314332f9ef05f5fa8168dfff174f6cfL31-R69), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-83b79888786fd47915f01e3ae6fe2d2fb314332f9ef05f5fa8168dfff174f6cfR119-R175), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-83b79888786fd47915f01e3ae6fe2d2fb314332f9ef05f5fa8168dfff174f6cfL87-R186), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-83b79888786fd47915f01e3ae6fe2d2fb314332f9ef05f5fa8168dfff174f6cfL94-R197), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-ba2503e087a690bc3fe435cec6567b05184eb244c86b94daa0bbe88e6ff66fecR1-R23), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-13cfae3c4d5e9cd31a80886624110c128b2041d46d79e175a95d672f90630faeR1-R23))
*  Add prevStepId field to Event type, table, and model to store the id of the previous stepBlock for the StepPrevEvent type ([link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-e60078ed0fc71b8cd3bb2fca25505e7dba866b7d463147662a12432dfe287273R1662-R1701), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3414f765c008fb078ca2038a887255fce5b7b9c1549d2b945e5e60663670dddaR1-R2), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-881378863f13d5e9573c2c2e756ae61e8161918ebb4a384a8d5dddd7bf7385b0R103), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-22bb985c2feeff20469c36cbf3e796ca0b36c367831d258a61cc42a52a5c5dc0R152), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-193cdc12303dd98cee8816df88d347a414164a780edfcbb4cf67527eeeb8fa31R243), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-ea8c85f7941c8cafb40fe8da944e8673862b9eb3c390bf99c32741559d4d33d7L39-R39), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-1c6df4c10d8c44ecc41187529f1381bb1cfdf7cedc7c7730646961e76e67d7b0L13-R13))
*  Add stepNextEventCreate mutation field to create and track events when users navigate to the next step in a journey ([link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-684ea1b2346c27f9252ea686988ffaa756a41f56bd7c62530441bfe2d034df3dR1), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-684ea1b2346c27f9252ea686988ffaa756a41f56bd7c62530441bfe2d034df3dL8-R21), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-684ea1b2346c27f9252ea686988ffaa756a41f56bd7c62530441bfe2d034df3dL35-R56), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-684ea1b2346c27f9252ea686988ffaa756a41f56bd7c62530441bfe2d034df3dL73-R154), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-98c806310428fde902045482ce23027cd6ab7578d9e4048de778be28eed4b4caR200-R215), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-83b79888786fd47915f01e3ae6fe2d2fb314332f9ef05f5fa8168dfff174f6cfL1-R11), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-83b79888786fd47915f01e3ae6fe2d2fb314332f9ef05f5fa8168dfff174f6cfL16-R41), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-83b79888786fd47915f01e3ae6fe2d2fb314332f9ef05f5fa8168dfff174f6cfL31-R69), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-83b79888786fd47915f01e3ae6fe2d2fb314332f9ef05f5fa8168dfff174f6cfR119-R175), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-83b79888786fd47915f01e3ae6fe2d2fb314332f9ef05f5fa8168dfff174f6cfL87-R186), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-83b79888786fd47915f01e3ae6fe2d2fb314332f9ef05f5fa8168dfff174f6cfL94-R197), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-ba2503e087a690bc3fe435cec6567b05184eb244c86b94daa0bbe88e6ff66fecR1-R23))
*  Fix typo in comment for nextStepId field of StepViewEvent type ([link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-e60078ed0fc71b8cd3bb2fca25505e7dba866b7d463147662a12432dfe287273L1647-R1652), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3cbd0d77b8eff95fb0f62db7f9f1a4b69ccd568ae3902a3450bac7d752c817a3L1310-R1317), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-29131681a98f41397b30e90886c80b6fc28d42df8c3c5b7df71856a044d9b5abL38-R46))
*  Add comment to blockId field of StepViewEvent type ([link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-e60078ed0fc71b8cd3bb2fca25505e7dba866b7d463147662a12432dfe287273R1728-R1729), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3cbd0d77b8eff95fb0f62db7f9f1a4b69ccd568ae3902a3450bac7d752c817a3R1284-R1285), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-29131681a98f41397b30e90886c80b6fc28d42df8c3c5b7df71856a044d9b5abR6-R8))
*  Wrap Card and NavigationButton components with MockedProvider component in test cases to mock GraphQL queries and mutations ([link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL1-R41), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL43-R130), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL60-R149), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beR161-R210), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL78-R219), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL89-R232), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL102-R247), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL112-R259), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL121-R270), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL131-R284), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL148-R303), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL159-R316), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL172-R331), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL182-R343), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL191-R354), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-3c3c31bf2041808306ffab49b753614c03f532376e6c0056aadb888efd19b2beL201-R366), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-031dbee4834f89ee19bea387a9766835e8c77e6075937db6eeb8aad8d6890ceeR1), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-031dbee4834f89ee19bea387a9766835e8c77e6075937db6eeb8aad8d6890ceeL114-R119), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-031dbee4834f89ee19bea387a9766835e8c77e6075937db6eeb8aad8d6890ceeL127-R139), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-031dbee4834f89ee19bea387a9766835e8c77e6075937db6eeb8aad8d6890ceeL143-R152), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-031dbee4834f89ee19bea387a9766835e8c77e6075937db6eeb8aad8d6890ceeL153-R164), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-031dbee4834f89ee19bea387a9766835e8c77e6075937db6eeb8aad8d6890ceeL163-R180), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-031dbee4834f89ee19bea387a9766835e8c77e6075937db6eeb8aad8d6890ceeL180-R198), [link](https://github.com/JesusFilm/core/pull/2027/files?diff=unified&w=0#diff-031dbee4834f89ee19bea387a9766835e8c77e6075937db6eeb8aad8d6890ceeL198-R218))

### CODE REVIEW

File: apps/api-journeys/schema.graphql [hunk 1]
```diff
@@ -1335,6 +1340,41 @@ type StepNextEvent implements Event {
   value: String
 }
 
+input StepPrevEventCreateInput {
```
Comment (csiyang) [inline apps/api-journeys/schema.graphql:1343]: All API changes are contained in https://github.com/JesusFilm/core/pull/2022

File: libs/journeys/ui/src/components/Card/Card.tsx [hunk 2]
```diff
@@ -28,8 +51,22 @@ export function Card({
   fullscreen,
   wrappers
 }: CardProps): ReactElement {
+  const [stepNextEventCreate] = useMutation<StepNextEventCreate>(
+    STEP_NEXT_EVENT_CREATE
+  )
+  const [stepPrevEventCreate] = useMutation<StepPreviousEventCreate>(
```
Comment (csiyang) [inline libs/journeys/ui/src/components/Card/Card.tsx:57]:
```suggestion
  const [stepPreviousEventCreate] = useMutation<StepPreviousEventCreate>(
```

File: apps/journeys/src/components/Conductor/NavigationButton/NavigationButton.tsx [hunk 3]
```diff
@@ -1,13 +1,24 @@
...
 import { ReactElement, useEffect } from 'react'
+import TagManager from 'react-gtm-module'
+import { useTranslation } from 'react-i18next'
+import { v4 as uuidv4 } from 'uuid'
 
 import type { TreeBlock } from '@core/journeys/ui/block'
 import { useBlocks } from '@core/journeys/ui/block'
+import { StepNextEventCreate } from '@core/journeys/ui/Card/__generated__/StepNextEventCreate'
+import { StepPreviousEventCreate } from '@core/journeys/ui/Card/__generated__/StepPreviousEventCreate'
+import {
```
Comment (tataihono) [inline apps/journeys/src/components/Conductor/NavigationButton/NavigationButton.tsx:17]: import should be from the local project.

File: apps/journeys/src/components/Conductor/NavigationButton/NavigationButton.tsx [hunk 4]
```diff
@@ -70,11 +89,70 @@ export function NavigationButton({
...
+        event: target === 'next' ? 'step_next' : 'step_prev',
+        eventId: id,
+        blockId: activeBlock.id,
+        stepName,
+        targetStepId: targetBlock?.id,
+        targetStepName
+      }
+    })
+  }
+
```
Comment (tataihono) [inline apps/journeys/src/components/Conductor/NavigationButton/NavigationButton.tsx:148]: this method is complex. split it into two. handleNavigationNextEvent etc.

File: libs/journeys/ui/src/components/Card/Card.tsx [hunk 5]
```diff
@@ -79,21 +116,86 @@ export function Card({
       (child) => child.__typename === 'VideoBlock' && child.id !== coverBlockId
     ) != null
 
+  // should always be called with nextActiveBlock() and previousActiveBlock()
+  // should match with other handleNavigationEventCreate functions
+  function handleNavigationEventCreate(target: 'next' | 'prev'): void {
```
Comment (tataihono) [inline libs/journeys/ui/src/components/Card/Card.tsx:121]: split into next and prev

File: libs/journeys/ui/src/components/Card/Card.tsx [hunk 6]
```diff
@@ -79,21 +116,86 @@ export function Card({
       (child) => child.__typename === 'VideoBlock' && child.id !== coverBlockId
     ) != null
 
+  // should always be called with nextActiveBlock() and previousActiveBlock()
+  // should match with other handleNavigationEventCreate functions
```
Comment (tataihono) [inline libs/journeys/ui/src/components/Card/Card.tsx:120]: you should reference the file to be helpful

**Review by tataihono — CHANGES_REQUESTED** [review]

**Review by tataihono — DISMISSED** [review]

**Review by tataihono — APPROVED** [review]

**Pull request merged**